### PR TITLE
[Profiler] fud2 support and processing instrumentation-produced cells

### PR DESCRIPTION
**Pull request by ayakayorihiro**
*Note: This PR is still a draft because I have some questions about the changes I made here that I wanted to ask about. Once those concerns are resolved I will turn this into a full PR.*
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
## This PR contains:                                                                                                                                                                                      
- `fud2/scripts/profiler.rhai`: fud2 support for the profiler. I added a new state `flamegraph` (the `svg` file containing the flame graph) and a new operation called  `profiler`, which takes a Calyx file and produces the profiled flame graph.
- Updates to existing profiling scripts to read instrumented cell signals and allow toggling between profiling optimized and non-optimized Calyx programs.
- Updated fud2 tests to reflect new state and operation.

## Usage

First, clone https://github.com/brendangregg/FlameGraph and edit the `fud2` configuration file to specify the location of `flamegraph.pl`:
ex)
```
[flamegraph]
script = "/home/ayaka/projects/FlameGraph/flamegraph.pl"
```

To obtain a flame graph from a Calyx program, specify the output as a `svg` file.
ex) 
```
fud2 tests/correctness/while.futil -o while.svg -s sim.data=tests/correctness/while.futil
```
will produce the below flame graph.
![image](https://github.com/user-attachments/assets/b257ecb6-9f45-49b7-9555-f68e3212403e)

To obtain a flame graph for the non-compiler-optimized version of the Calyx program, add `-s passes=no-opt`:
ex)
```
fud2 tests/correctness/while.futil -o flame.svg -s sim.data=tests/correctness/while.futil.data -s passes=no-opt
```

### CODE REVIEW

**Comment by rachitnigam** [comment]
Exciting stuff! A couple of questions:
> A tool to produce JSON listing all groups names for each component (very similar to tools/component_cells). This was necessary because the VCD parsing script needs to know what groups it should profile. (I previously got this list of groups from TDCC, but the profiler needs a way to discover groups at the source level that were optimized away)

Is this something the `resources` backend could support? Would be nice to fold in that functionality into one location instead of having a new tool.

> First, clone https://github.com/brendangregg/FlameGraph and edit the fud2 configuration file to specify the location of flamegraph.pl

Is there a way to install this from `brew` or something else as well? Is that a better default?

> This fud2 pipeline basically does what the tools/profiler/get-profile-counts-info.sh does. Should I remove the bash script?

Yes! The less bash the better IMO!

> I later realized I might be able to work around the problem that required me to make the component_groups tool. Should I implement that workaround in this PR (and get rid of component_groups) or do that in the future?

If it's a small change then please remove the tool. Otherwise a follow-up PR sounds good!

File: calyx-backend/src/verilog.rs [hunk 1]
```diff
@@ -539,8 +539,8 @@ fn emit_assignment(
 
     // If this is a data port
     let rhs: v::Expr = if is_data_port(dst) {
-        if assignments.len() == 1 {
```
Comment (ekiwi) [inline calyx-backend/src/verilog.rs:542]: Could you factor the Verilog backend changes into a separate PR? I am not sure why we would need this.
It might be possible to just mark the `cell` that you create in your instrumentation pass with `@control` which would be a cleaner solution.

Comment (rachitnigam) [reply calyx-backend/src/verilog.rs:542]: Good catch. Instrumentation should not change the backend!

Comment (ayakayorihiro) [reply calyx-backend/src/verilog.rs:542]: I'll try looking into this! I think the idea here is that we get the instrumentation pass to override `std_wire`'s `@data` input?

Comment (rachitnigam) [reply calyx-backend/src/verilog.rs:542]: Can you describe the interaction between `@data` and `@protected`? Why is something being marked as `@protected` change its data semantics?

Comment (ayakayorihiro) [reply calyx-backend/src/verilog.rs]: Marking the cell as `@control` fixed the problem so I removed the Verilog backend changes!

Comment (rachitnigam) [reply calyx-backend/src/verilog.rs]: So the instrumentation generates protected control signal? What is the rationale for these signals being control?

I ask because if the semantics of control change in the future, it will silently break instrumentation if we dont have a precise reasoning for why these signals are marked control.

File: calyx-opt/src/passes/profiler_instrumentation.rs [hunk 2]
```diff
@@ -52,10 +52,12 @@ impl Visitor for ProfilerInstrumentation {
             for group_name in group_names.into_iter() {
                 let name = format!("{}_probe", group_name);
                 let inst_cell = builder.add_primitive(name, "std_wire", &[1]);
+                // inst_cell.borrow_mut().attributes
```
[the remark this answers is not in the document]
Comment (ayakayorihiro) [reply calyx-opt/src/passes/profiler_instrumentation.rs]: Fixed! Thanks for the catch.

File: calyx-backend/src/verilog.rs (hunk 1 above)
Comment (ayakayorihiro) [reply calyx-backend/src/verilog.rs]: That's fair - the problem is that I'm trying to use `std_wire` for instrumented cells, but I can't have the Verilog backend reduce the assignment to `<group>_probe_in = 1'd1` (which it will if I don't make the cells `@control`, because the input port of `std_wire` is `@data`).

I think there are a couple of paths that we can take, most of which I've explored so far: (1) have a group that initializes all `<group>_probe.in` to `0` (which is not ideal because we'd need extra cycles to execute this group first) (2) make a new primitive that doesn't have a `@data` input port, (3) have the instrumentation pass make its cells `@control` (which is my current changes here), and (4) have `@protected` mean that the Verilog backend should preserve the original assignment (my original changes here).
If you have other suggestions though, I'd be happy to hear them!

**Comment by ayakayorihiro** [comment]
Thanks for the questions! Answering them here: 

> Is this something the resources backend could support? Would be nice to fold in that functionality into one location instead of having a new tool.

The resources backend gives you the list of primitives, so this tool would be giving us the complement (non-primitives). But I think I'm leaning towards getting rid of the tool and instead having the `profiler-instrumentation` pass give me information about the groups that it instrumented. I think it should be a small enough change that I can include it in this PR!

> Is there a way to install this from brew or something else as well? Is that a better default?

I unfortunately only have a Mac at home, so I'll check this out tonight. But there doesn't seem to be a Linux/Windows installation command...

Also, I removed the bash script!

File: calyx-backend/src/verilog.rs (hunk 1 above)
Comment (rachitnigam) [reply calyx-backend/src/verilog.rs]: Can you explain how the probes work? It seems you want them to be set to `1` when a particular group is executing and otherwise have them be `0`? If so, the `@control` annotation makes sense to me since it provides this precise guarantee: when there is an explicit assignment, it takes that value; otherwise, it asserts `0`.

Comment (ayakayorihiro) [reply calyx-backend/src/verilog.rs]: Yup that's exactly how the probes work!! I'm glad to hear that makes sense - I'll write some comments documenting the reason why we want it to be a `@control` cell.

File: fud2/scripts/profiler.rhai [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    // series of passes after instrumentation?
+    e.config_var_or("passes", "passes", "all"); // set passes="no-opt" to run without optimizations
+
+    // rules for preprocessing
+
+    e.config_var_or("component_cells", "component_cells", "$calyx-base/target/debug/component_cells");
+    e.rule("component-cells", "$component_cells -l $calyx-base $in > $out");
+
+    // dummy rule to force ninja to build the testbench
+    e.rule("dummy", "sh -c 'cat $$0' $in > $out");
```
[the remark this answers is not in the document]
Comment (ayakayorihiro) [reply fud2/scripts/profiler.rhai:23]: Thanks for the catch, I don't need it anymore!

File: tools/profiler/create-visuals.py [hunk 4]
```diff
@@ -41,7 +41,13 @@ def get_active_group(self):
         elif len(self.active_groups) == 1:
             return self.active_groups[0]
         else:
-            raise Exception(f'Component {self.component} is parallel! Active groups: {self.active_groups}')
+            # concatenate all parallel active groups
+            sorted_active_groups = list(sorted(self.active_groups))
+            acc = sorted_active_groups[0]
+            for i in range(1, len(sorted_active_groups)):
```
[the remark this answers is not in the document]
Comment (ayakayorihiro) [reply tools/profiler/create-visuals.py:47]: That makes so much sense, I just went with what I remembered from intro CS!
